src/main/java/com/aliyun/odps/jdbc/OdpsPreparedStatement.java

```diff
@@ -42,27 +45,71 @@
 import java.sql.SQLFeatureNotSupportedException;
 import java.sql.Types;
 import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Calendar;
 import java.util.HashMap;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.aliyun.odps.OdpsType;
+import com.aliyun.odps.TableSchema;
+import com.aliyun.odps.data.Record;
+import com.aliyun.odps.tunnel.TableTunnel;
+import com.aliyun.odps.tunnel.TunnelException;
+import com.aliyun.odps.tunnel.io.TunnelRecordWriter;
 
 public class OdpsPreparedStatement extends OdpsStatement implements PreparedStatement {
 
+  private final String TABLE_NAME = "((\\w+\\.)?\\w+)";      // "proj.name" or "name"
+  private final String PREP_VALUES = "\\([\\?,?\\s*]+\\)"; // "(?, ?, ?)" or "(?)"
+  private final String SPEC_COLS = "\\([\\w+,?\\s*]+\\)";  // "(name1, name2, name3)"
+
+  private final String PREP_INSERT_WITH_SPEC_COLS =
+      "(?i)^\\s*insert\\s+into\\s+" + TABLE_NAME + "\\s+" + SPEC_COLS + "\\s+values\\s+"
+      + PREP_VALUES + "\\s*;?\\s*$";
+  private final String PREP_INSERT_WITH_SPEC_COLS_EXAMPLE =
+      "INSERT INTO table (field1, field2) VALUES (?, ?);";
+
+  private final String PREP_INSERT_WITHOUT_SPEC_COLS =
+      "(?i)^\\s*insert\\s+into\\s+" + TABLE_NAME + "\\s+values\\s+" + PREP_VALUES + "\\s*;?\\s*$";
+  private final String
+      PREP_INSERT_WITHOUT_SPEC_COLS_EXAMPLE =
+      "INSERT INTO table VALUES (?, ?, ?);";
+
   /**
    * The prepared sql template (immutable).
    * e.g. insert into table FOO select * from BAR where id = ? and weight = ?
    */
   private final String sql;
 
+  private boolean verified = false;
+  private String tableBatchInsertTo;
+
+  private int parametersNum;
+
   /**
    * The parameters for the prepared sql (index=>parameter).
-   * The parameter is stored as a string literal value so that we can insert them
-   * to the prepared sql template directly.
+   * The parameter is stored as Java objects and lazily casted into String when submitting sql.
+   * The executeBatch() call will utilize it to upload data to ODPS via tunnel.
    */
-  private final HashMap<Integer, String> parameters = new HashMap<Integer, String>();
+  private HashMap<Integer, Object> parameters = new HashMap<Integer, Object>();
+
+  // When addBatch(), compress the parameters into a row
+  private List<Object[]> batchedRows = new ArrayList<Object[]>();
 
   OdpsPreparedStatement(OdpsConnection conn, String sql) {
     super(conn);
     this.sql = sql;
```

Comment:
Seems more reasonable to do verify logic here. I'm not sure if throw exception in constructor is a good idea. However, if not verified, it is meaningless for user to do any further action, such as setXXX and addBatch.